File: e-welink.ts
const MICROBIT_ID_EWELINK = 2301;
const MICROBIT_EWELINK_EVT_ON_WRITTEN = 60;

enum EWeLinkDirection {
    //% block="stop"
    Stop,
    //% block="left"
    Left,
    //% block="right"
    Right,
    //% block="up"
    Up,
    //% block="down"
    Down,
}

/**
 * Supports for eWeLink service.
 */
//% weight=50 icon="\uf1eb" color=#1976d2
namespace eWeLink {
    let onPressHandlers: { [direction: string]: () => void } = {};
    let pressDirection: EWeLinkDirection = EWeLinkDirection.Stop;

    /**
     * Starts service.
     */
    //% blockId=ewelink_start block="eWeLink start"
    //% weight=90 blockGap=8
    export function start(): void {
        startService();

        // on received
        control.onEvent(MICROBIT_ID_EWELINK, MICROBIT_EWELINK_EVT_ON_WRITTEN, () => {
            const buffer = readBuffer();
            if (buffer.length > 0) {
                const data = buffer.toArray(NumberFormat.UInt8LE);
                const sum = data.reduce((previous, current) => previous + current, 0);
                if (2 * data[data.length - 1] === sum) {
                    const cmd: number = data[4];
                    pressDirection = cmd;
                    onPressHandlers[cmd] && onPressHandlers[cmd]();
                }
            }
        });
    }

    /**
     * On eWeLink controller pressed.
     */
    //% blockId=ewelink_onpressed
    //% block="on eWeLink|%direction pressed"
    //% weight=70 blockGap=8
    export function onPressed(direction: EWeLinkDirection, handler: () => void) {
        onPressHandlers[direction] = handler;
    }

    /**
     * Get eWeLink controller press state.
     */
    //% blockId=ewelink_ispressed
    //% block="eWeLink|%direction is pressed"
    //% weight=60 blockGap=32
    export function isPressed(direction: EWeLinkDirection) {
        return direction === pressDirection;
    }

    /**
     * Returns true if the service is connected.
     */
    //% blockId=ewelink_isconnected block="eWeLink is connected"
    //% weight=20 blockGap=8 shim=eWeLink::isConnected
    export function isConnected(): boolean {
        return false;
    }

    /**
     * Sets the bluetooth transmit power between 0 (minimal) and 7 (maximum).
     * @param power power level between 0 (minimal) and 7 (maximum), eg: 7.
     */
    //% weight=10
    //% blockId=ewelink_settransmitpower block="eWeLink|set transmit power %power"
    //% power.min=0 power.max=7 shim=eWeLink::setTransmitPower
    //% blockGap=8
    export function setTransmitPower(power: number) {
        return;
    }

    /**
     * Starts the eWeLink service.
     */
    //% shim=eWeLink::startService
    export function startService(): void {
        return;
    }

    /**
     * read buffer.
     */
    //% shim=eWeLink::readBuffer
    export function readBuffer(): Buffer {
        return Buffer.create(0);
    }
}

File: e-welink.cpp
#include "pxt.h"
#include "MicroBit.h"
#include "MicroBitConfig.h"
#include "EWeLinkService.h"

//% weight=50 icon="\uf1eb" color=#1976d2
namespace eWeLink
{
    EWeLinkService *pService = NULL;
    bool connected = false;
    bool nameScrolling = true;

    // Creates a connected image.
    MicroBitImage connectedImage("\
        000,000,000,000,000\n\
        000,000,000,000,255\n\
        000,000,000,255,000\n\
        255,000,255,000,000\n\
        000,255,000,000,000\n");

    void onConnected(MicroBitEvent)
    {
        nameScrolling = false;
        uBit.display.stopAnimation();
        uBit.display.print(connectedImage, 0, 0, 0, 888);
        uBit.display.clear();
        connected = true;
    }

    void onDisconnected(MicroBitEvent)
    {
        uBit.reset();
    }

    void scrollFriendlyName()
    {
        while (nameScrolling)
        {
            uBit.display.scroll(microbit_friendly_name());
            uBit.sleep(1000);
        }
    }

    //%
    void startService()
    {
        if (pService != NULL)
            return;

        uBit.messageBus.listen(MICROBIT_ID_BLE, MICROBIT_BLE_EVT_CONNECTED, onConnected);
        uBit.messageBus.listen(MICROBIT_ID_BLE, MICROBIT_BLE_EVT_DISCONNECTED, onDisconnected);

        pService = new EWeLinkService(*uBit.ble);

        create_fiber(scrollFriendlyName);
    }

    //%
    bool isConnected()
    {
        return connected;
    }

    //%
    void setTransmitPower(int power)
    {
        uBit.bleManager.setTransmitPower(min(MICROBIT_BLE_POWER_LEVELS - 1, max(0, power)));
    }

    //%
    Buffer readBuffer()
    {
        auto buffer = mkBuffer(NULL, EWELINK_DATA_LENGTH);
        auto res = buffer;
        registerGCObj(buffer);
        uint8_t len = pService->read(buffer->data);
        if (len == 0)
        {
            res = mkBuffer(NULL, 0);
        }
        else
        {
            res = mkBuffer(buffer->data, len);
        }
        unregisterGCObj(buffer);
        return res;
    }
}
